plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -447,6 +450,21 @@ void logTimings() {
         }
     }
 
+    /**
+     * Mark a call to an internal API made by this build.
+     * @param call a representation of the call site; for example, {@code hudson.model.Run.setDescription}
+     */
+    synchronized void recordInternalCall(@NonNull String call) {
+        if (internalCalls == null) {
+            internalCalls = new TreeSet<>();
```

Comment:
Given that we have custom serialization logic here anyway, would it make sense to use `ConcurrentSkipListSet` as the in-memory type to avoid having to synchronize on `CpsFlowExecution` for all writes?